File: haskell/space-age/src/SpaceAge.hs
module SpaceAge (Planet(..), ageOn) where

data Planet = Mercury
            | Venus
            | Earth
            | Mars
            | Jupiter
            | Saturn
            | Uranus
            | Neptune

earthSecondsAYear :: Float
earthSecondsAYear = 31557600

ageOn :: Planet -> Float -> Float
ageOn planet second = second / factor planet / earthSecondsAYear

factor :: Planet -> Float
factor Mercury = 0.2408467
factor Venus = 0.61519726
factor Earth = 1
factor Mars = 1.8808158
factor Jupiter = 11.862615
factor Saturn = 29.447498
factor Uranus = 84.016846
factor Neptune = 164.79132

File: c/space-age/space_age.c
#include "space_age.h"

float age(const planet_t planet, const int64_t seconds) {
  float ratio;
  switch (planet) {
  case MERCURY:
    ratio = EARTH_YEARS_MERCURY;
    break;
  case VENUS:
    ratio = EARTH_YEARS_VENUS;
    break;
  case EARTH:
    ratio = EARTH_YEARS_EARTH;
    break;
  case MARS:
    ratio = EARTH_YEARS_MARS;
    break;
  case JUPITER:
    ratio = EARTH_YEARS_JUPITER;
    break;
  case SATURN:
    ratio = EARTH_YEARS_SATURN;
    break;
  case URANUS:
    ratio = EARTH_YEARS_URANUS;
    break;
  case NEPTUNE:
    ratio = EARTH_YEARS_NEPTUNE;
    break;
  default:
    return -1;
  }
  const float earth_years = (float)seconds / EARTH_SECONDS_TO_YEARS;
  return earth_years / ratio;
}
